### ISSUE TITLE
Standardize Comment Types

### ISSUE THREAD

**Issue by GuyLewin** [estree/estree#201]
Acorn and Esprima both create comments with types "Block" or "Line".
Meriyah (and Cherow) create comments with types "SingleLine", "MultiLine", "HashbangComment", "HTMLOpen" or "HTMLClose".
I love Meriyah's support of different single-line comments, and I think these 5 types should be standardized.

**Comment by loganfsmyth** [comment]
I think the harder part of this would be defining what comments go into which nodes. If I write
```
function foo() /* comment */ { console.log(""); }
```
`comment` is a `BlockComment`, but what AST node should that be attached to? In my mind, that's the motivation for not including comments in this spec. Since this isn't a CST, comments are ambiguous.

**Comment by GuyLewin** [comment]
@loganfsmyth I agree comments are ambiguous, but right now there's a mess in parsing and and attaching them. 
Can't we standartize the fact that every comment like in your example has to go to either the 1st node's trailingComments or the 2nd node's leadingComments? Including their types. 
I believe that even just these claims will make the libraries work together much better than they do now.

**Comment by kaicataldo** [comment]
For what it's worth, we've moved away from comment attachment in ESLint due to the challenges above and we now process comments in the context of token lists.

**Comment by nzakas** [comment]
While I think the idea of standardizing this has merit, in practice existing parsers won’t be able to make any changes due to compatibility issues. ESLint, for example, couldn’t make a change to the nodes we use because there is already a large ecosystem depending on what we already produce. (Not to mention compatible parsers like babel-eslint that would also need to change.)

**Comment by brettz9** [comment]
@nzakas : Could you clarify whether you are concerned:

1. That the standard would treat Comments as fully integrated members of the tree, e.g., as independent children of say `Program.body`
2. That even just adding a property to other nodes, e.g., adding a `comments` property to `Program`, `VariableDeclaration`, etc., would be problematic for compatibility as far as diffs or potential conflicts with tools which are already extending the AST in such a manner?

If the concern includes the latter, it seems that at least for your espree parser, the [README](https://github.com/eslint/espree) seems to leave open that exact comment structuring might change:

> ...Espree may deviate from what Esprima outputs in terms of where and how comments are attached, as well as what additional information is available on AST nodes. That is to say, Espree may add more things to the AST nodes than Esprima does but the overall AST structure produced will be the same.

I am therefore wondering if only no. 1 is a real concern here.

If so, I'd also be interested in discussing whether, regardless of how they may be attached to nodes, a standard AST structure could be settled on for jsdoc-style comments (along the lines of the output produced by `comment-parser` or your now abandoned `doctrine`).

While the eslint organization may not be specifically interested in maintaining comment-aware rules, there are many use cases, from linting to querying, for such comments. It seems to me that jsdoc is already a virtual and seemingly lone structured comment standard, so it is not like this would be introducing something in competition with other popular comment structures (even if provision would need to be made for non-jsdoc comments).

There would admittedly be a question of exactly how to represent jsdoc comments, e.g., whether in a manner that is at least somewhat aware of tags (e.g., at least that well-known tags such as `@returns` should not be parsed for a name while `@param` should be) or which is wholly generic (e.g., parsing of `@returns My description` would give `My` as the "name" of the tag) and though this might not satisfy everyone, regardless of the form chosen, having some structure ought to help.

**Comment by kaicataldo** [comment]
@brettz9 So I understand, are you suggesting we add JSDoc comments as a defined node in ESTree?

Full disclosure: I worked on removing comment attachment from Espree/ESLint. 

Generally, I don't think comments in the context of an AST make sense. Since they can occur between any two tokens, figuring out which nodes the comments should be attached to is guesswork at best.

While I do think it would be great if the JS community had a standardized JSDoc spec/parser (which I think would be possible using either JSDoc or TypeScript's comment parsing logic, though I don't think they're exposed publicly), I don't think it belongs in ESTree. Contexts in which comments have special meaning are specific to their respective environments (JSDoc, webpack, Uglify, etc.), and finding these comments relative to the nodes they are "attached" to in these specific contexts is a lot easier than trying to guess where all the comments in a program should be attached. I've personally found it a lot easier/more accurate to locate comments by iterating over tokens rather than relying on comment attachment at the parser level.

**Comment by kaicataldo** [comment]
@GuyLewin To your original point, agreed that it would be nice to standardize these type names, but I share @nzakas's concern about causing pain in the ecosystem.

**Comment by brettz9** [comment]
> So I understand, are you suggesting we add JSDoc comments as a defined node type in ESTree?

Not as an independent Node (e.g., one with a `type` and within the `body` of other nodes), as I can see that would be backward-breaking and also not of interest to all users. But as structured meta-data, as some parsers have a `comments` property attached to Nodes.

> Generally, I don't think comments in the context of an AST make sense. Since they can occur between any two tokens, figuring out which nodes the comments should be attached to is guesswork at best.

Is it not possible to settle on a few basic rules, e.g., that it should always be attached to the following node (though with inline comments following if they are also being represented)? While an *abstract* tree of course ought to aim for semantic soundness to its groupings, I think in the case of such meta-data, the important thing is being able to get at it in some consistent manner. While the resulting structure must be determined by settling on some arbitrary rules, when standardized, the arbitrariness becomes irrelevant as it becomes understood how to find one's way around (e.g.,. to look for comments on the Node that follows them).

> While I do think it would be great if the JS community had a standardized JSDoc spec/parser (which I think would be possible using either JSDoc or TypeScript's comment parsing logic, though I don't think they're exposed publicly), 

Generic comment parsing, as with TypeScript, appears to be integrated along with their generic parsers, and not so easy to get at, even if possible through the public APIs. For JSDoc's type parsing, the case is more clear, as it is using its own `catharsis` library. (This comment may be of interest re: TS and type parsing: https://github.com/jsdoctypeparser/jsdoctypeparser/issues/109#issuecomment-570708263 )

> Contexts in which comments have special meaning are specific to their respective environments (JSDoc, webpack, Uglify, etc.), and finding these comments relative to the nodes they are "attached" to in these specific contexts is a lot easier than trying to guess where all the comments in a program should be attached. I've personally found it a lot easier/more accurate to locate comments by iterating over tokens rather than relying on comment attachment at the parser level.

But isn't that a chicken-and-egg problem that because there is no standard, there is therefore no reliable way to know how to find attached comments when a parser exposes them? If a standard were adopted to always add them to the following Node (or to `Program` when there is only a comment in the doc) or by some other justifiable approach, there would be no need to re-retrieve/relocate tokens, and it would become clear and could be accurate (and more efficient). Also I think JSDoc is a lot more general purpose than the others.

And although JSDoc does have some of its own semantics (specific tags with specific meaning), it is more of a syntax (and as mentioned, the resulting structure could be agnostic as to vocabulary like [`comment-parser`](https://github.com/yavorskiy/comment-parser) which parses jsdoc or jsdoc-like syntax agnostically as to the meaning of the tags, e.g., all blocks could be summarized as:

```
/**
@aTag {optionalTypes} [aName=anyDefaults] Description
@anotherTag {optionalTypes} [aName=anyDefaults] Description
*/
```

While yes, there are other uses to comments out there, none to my knowledge use the specific jsdoc syntax `/** ... */` with an incompatible internal structure.

In summary, if the decision is against out-of-the-box pre-parsed jsdoc, I think it would be nice to at least get at the comments in a reliable and efficient manner by looking on a node of interest (or its parent or child), assuming some standardized choice.

**Comment by bradzacher** [comment]
My 2c - comment parsing does not belong encoded in the AST for two reasons:
Comments have no bearing on the operation of a piece of JS code.
Comments have no specified standard position wherein they have a specific meaning, without accounting for the context within which they are consumed, meaning ESTree would have to pick a comment spec to support, and go with that.

Attaching them is an inexact science, and there's no hard-and-fast rule which applies. As @kaicataldo mentioned - the way in which a comment attach to a node is entirely determined by the context in which the comment is consumed, if it even matters at all.

For example, typescript supports JSDoc comments, and flow supports flow type comments, so the rule of "attach to the following node" doesn't always hold true, and you have to use the context within which you are analysing the comments to determine how you should attach the comment

```js
/** @type {number} */ const tsTyped = something();
const flowTyped /*: number*/ = something();

/** @returns {number} */
function tsReturnTyped() {}
function flowReturnTyped() /*: number*/ {}

function tsArgs(/** @type {number} */ arg, /** @type {string} */ arg2) {}
function flowArgs(arg /*: number */, arg2 /*: string */) {}
```

Then there is the problem as mentioned of misc inline comments, how are these handled? Where are they attached in a representative way? Or are they just special case ignored?

If they're ignored, won't that cause problems with comment analysers that care about inline comments?
If they're not ignored, how do you encode the following into an AST?

```js
/* */ async /* */ function /* */ * /* */ foo /* */ ( /* */ arg /* */, /* */ ) /* */ { }

/* */ const /* */ x /* */ = /* */ { /* */ prop /* */ : /* */ 1 /* */ } /* */;

/* */ return /* */ await /* */ foo /* */ ( /* */ arg /* */ ) /* */;
```

You can, but you'd have to go through every single node, and analyse every single position that a block or `//` comment is allowed, and then encode that into the AST, which is no small feat.
And you have to define a standard around when a comment attaches to node X and when it attaches to node Y.

I'm sure that there are different edge cases which change where the comment attaches to in certain cases (just like we have contextual keywords like `await` - `await(1);` is a call expression in a global context, but is an await expression in an async function context).

All of these are solvable problems, for sure; with a lot of work and consensus it is solvable.
But my question is - what do you gain that you can't already do via methods like ESLint presents 
- analysing comments in the context of tokens?

**Comment by nzakas** [comment]
My main point is that ESTree is not actually a standard — it’s an agreement between a bunch of OSS packages to do things a certain way. Because of that, it’s just not possible to *change* what is required when there’s an ecosystem depending on the way it’s already been done.

I’m also of the mind that comments should not be in an AST for all the reasons others have mentioned.

While I admire all attempts to standardize things, it’s just too late in ESTree’s lifecycle for something like this to happen.

**Comment by brettz9** [comment]
> Comments have no bearing on the operation of a piece of JS code.

Sometimes they can:

```js
eval(('' + (_ => {/* alert('Hi') */})).slice(9, -3))
```

(Btw of possible side interest to this discussion: https://github.com/tc39/proposal-binary-ast/issues/76 )

> Comments have no specified standard position wherein they have a specific meaning, without accounting for the context within which they are consumed, meaning ESTree would have to pick a comment spec to support, and go with that.

ESTree would not need to bless a particular usage of comments. Only attaching in some consistent manner. Yes, you do effectively highlight some of the complexities of inline comments, as the tokens those comments surround are not necessarily full-blown nodes, but as you suggest this could be worked around. Different specs might need to look to a preceding or subsequent node or parent to find the comment of particular interest to them.

> For example, typescript supports JSDoc comments, and flow supports flow type comments, so the rule of "attach to the following node" doesn't always hold true, and you have to use the context within which you are analysing the comments to determine how you should attach the comment

Regarding ""attach to the following node" doesn't always hold true", it does hold true if the meta-data was added to the AST in a consistent manner , it's just that tools for JSDoc, Flow, etc. would need to check before or after to find the location suitable to their needs. But it would not require iterating through the whole document again to find them.

> what do you gain that you can't already do via methods like ESLint presents - surfacing comments in the context of tokens they are near to, and the tooling then using its encoded context to attach them itself?

Besides being a useful aid when say logging the nodes, I would expect you should be able to get better performance.  Below I trace the behavior in eslint and how it has to iterate through all of the comments in a document just to find comments adjacent to a Node.

@nzakas : Are you saying things would break just by adding `comments` (or some other property if conflicting) as meta-data? Why would the ecosystem have problems with an extra  meta-data property? Is diffing the issue? Standards can be developed working around de facto behavior, and I'm just not clear that as long as comments are only added as meta-data on adjacent nodes, and not inserted as co-equal Nodes of that tree, that such cannot be done backward-compatibly. The concerns with unique uses of comments do not argue against inclusion in the AST, as far as I see it, but rather argue against the AST suggesting the attachment of comment meta-data blesses any particular usage or that its attachment is anything more than a convenience.

----
**Only if of interest:**

To take one such method ESLint presents as an example, [getCommentsBefore](https://eslint.org/docs/developer-guide/working-with-rules#context-getsourcecode), looking at its [definition](https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/source-code/token-store/index.js#L583), we see it calls `createCursorWithCount` which passes on the Node information, the comments*, and a cursor, [`cursors.backward`](https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/source-code/token-store/cursors.js#L90) as its `CursorFactory`, with the 2nd argument to its constructor, `TokenCommentCursor`, being of most interest, and in this case being `BackwardTokenCommentCursor`. This cursor extends from `Cursor` which has a `getOneToken` method which calls the `moveNext` method (overridden by [BackwardTokenCommentCursor](https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/source-code/token-store/backward-token-comment-cursor.js#L41)). This uses `commentIndex` to find its place which [was obtained](https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/source-code/token-store/backward-token-comment-cursor.js#L36) by a [search](https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/source-code/token-store/utils.js#L38) which iterates through the comment tokens passed to it, in this case, being all comments of the source code.
 
\* For the comments that are passed, stepping back, we see [SourceCode](https://github.com/eslint/eslint/blob/33efd71d7c3496b4b9cbfe006280527064940826/lib/source-code/source-code.js#L147) extends [`TokenStore`](https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/source-code/token-store/index.js#L205) and calls the `TokenStore` super class with [`ast.comments` wholly independently from the `ast.tokens` it also passes](https://github.com/eslint/eslint/blob/33efd71d7c3496b4b9cbfe006280527064940826/lib/source-code/source-code.js#L175). So the `this[COMMENTS]` which are passed in by `getCommentsBefore` are all these comments.

**Comment by nzakas** [comment]
Just to close the loop on this, it’s not practical to try to standardize comment nodes as part of an AST. I’ve already explained why that is. I thank everyone for their thoughts and comments.
